### PR TITLE
New module: gitlab_milestone

### PR DESCRIPTION
**Pull request by gpongelli**
##### SUMMARY
A new Ansible module gitlab_milestone for creating, updating or deleting Gitlab project's and group's milestones.

##### ISSUE TYPE
- New Module/Plugin Pull Request

##### COMPONENT NAME
`gitlab_milestone`

##### ADDITIONAL INFORMATION
A new Ansible module `gitlab_milestone` for creating, updating or deleting GitLab project's and group's milestone resources utilizing [python-gitlab](https://python-gitlab.readthedocs.io/en/stable/) library. 

This work is based on previously available GitLab-related modules in the collection. 

Looking at the entire Ansible collection, there's no module that manipulates Gitlab project's and group's milestones.

### CODE REVIEW

File: plugins/modules/gitlab_milestone.py [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+
+# Copyright (c) 2023, Gabriele Pongelli (gabriele.pongelli@gmail.com)
+# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
+# SPDX-License-Identifier: GPL-3.0-or-later
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = '''
+module: gitlab_milestone
+short_description: Creates/updates/deletes GitLab Milestones belonging to project or group.
```
Comment (felixfontein) [inline plugins/modules/gitlab_milestone.py:12]:
```suggestion
short_description: Creates/updates/deletes GitLab Milestones belonging to project or group
version_added: 8.3.0
```

File: plugins/modules/gitlab_milestone.py [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+description:
+  - When a milestone does not exist, it will be created.
+  - When a milestone does exist, its value will be updated when the values are different.
+  - Milestones can be purged
+  - Task can be run in check mode
```
Comment (felixfontein) [inline plugins/modules/gitlab_milestone.py:17]:
```suggestion
  - Milestones can be purged.
```

File: plugins/modules/gitlab_milestone.py [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+    choices: ["present", "absent"]
+  purge:
+    description:
+      - When set to V(true), delete all milestone which are not mentioned in the task.
+    default: false
+    type: bool
+    required: false
+  project:
+    description:
+      - The path and name of the project, mandatory field but it's alternatively with group.
```
Comment (felixfontein) [inline plugins/modules/gitlab_milestone.py:50]:
```suggestion
      - The path and name of the project. Either this or O(group) is required.
```
Same below.

File: plugins/modules/gitlab_milestone.py [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+EXAMPLES = '''
+# same project's task can be executed for group
+- name: Create one milestone
+  community.general.gitlab_milestone:
+    api_url: https://gitlab.com
+    api_token: secret_access_token
+    project: "group1/project1"
+    milestones:
+      - title: milestone_one
+        start_date: 2024-01-04
```
Comment (felixfontein) [inline plugins/modules/gitlab_milestone.py:101]: This specifies a YAML date. Did you check that your code receives this as expected?

Comment (gpongelli) [reply plugins/modules/gitlab_milestone.py:101]: No, it’s expected to be a string with YYYY-MM-DD format. I’ll add double quotes there, thanks

Comment (gpongelli) [reply plugins/modules/gitlab_milestone.py]: Changed

File: plugins/modules/gitlab_milestone.py [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+      - The path and name of the project. Either this or O(group) is required.
+    required: false
+    type: str
+  group:
+    description:
+      - The path of the group, mandatory field but it's alternatively with project.
+    required: false
+    type: str
+  milestones:
+    version_added: 1.0.0
```
Comment (felixfontein) [inline plugins/modules/gitlab_milestone.py:58]:
```suggestion
```

File: plugins/modules/gitlab_milestone.py [hunk 6]
```diff
@@ -0,0 +1,498 @@
...
+  check_mode:
+    support: full
+  diff_mode:
+    support: none
+
+options:
+  state:
+    description:
+      - Create or delete milestone.
+      - Possible values are present and absent.
```
Comment (felixfontein) [inline plugins/modules/gitlab_milestone.py:37]:
```suggestion
```
The `choices` already says that.

File: plugins/modules/gitlab_milestone.py [hunk 7]
```diff
@@ -0,0 +1,498 @@
...
+    type: bool
+    required: false
+  project:
+    description:
+      - The path and name of the project. Either this or O(group) is required.
+    required: false
+    type: str
+  group:
+    description:
+      - The path of the group, mandatory field but it's alternatively with project.
```
Comment (felixfontein) [inline plugins/modules/gitlab_milestone.py:54]: This text also needs to be adjusted, similar to the one in `project`.

File: plugins/modules/gitlab_milestone.py [hunk 8]
```diff
@@ -0,0 +1,498 @@
...
+        default: null
+      start_date:
+        description:
+          - Milestone start date in YYYY-MM-DD format.
+        type: str
+        required: false
+        default: null
+      description:
+        description:
+          - Milestone's description
```
Comment (felixfontein) [inline plugins/modules/gitlab_milestone.py:85]:
```suggestion
          - Milestone's description.
```

File: plugins/modules/gitlab_milestone.py [hunk 9]
```diff
@@ -0,0 +1,498 @@
...
+    state: absent
+    milestones:
+      - title: milestone-abc123
+      - title: milestone-two
+'''
+
+RETURN = '''
+milestones:
+  description: Four lists of the milestones which were added, updated, removed or exist.
+  returned: always
```
Comment (felixfontein) [inline plugins/modules/gitlab_milestone.py:179]: I'm pretty sure this isn't returned when the module fails.
```suggestion
  returned: success
```
Same below.

**Comment by felixfontein** [comment]
@gpongelli thanks for your contribution!

**Pull request merged**